### PR TITLE
Remove name-mangling for almost all characters

### PR DESCRIPTION
**Pull request by soc**
While the list of exceptions hasn't changed much in size,
non-dangerous method names like + and ??? as well as 99% of the
character plane which were formerly name-mangled with $uXXXX appear
now as-is in stacktraces, other places and tools not specificially
written with Scala's name-mangling in mind.

The changes are mostly in line with the recommendations for the JVM
(look for "symbolic freedom"), except for the list of exceptions
which currently keep using the $-style mangling instead of the suggested
backslash-style (example: / is currently encoded as $slash, the
guidelines recommend |) due to readability concerns.

The improvements to readability are impressive, imho.
Another benefit is that people have to deal with a lot less mangled
names when working with macros.

### CODE REVIEW

**Comment by soc** [comment]
I created this one to gather opinions whether this is the right way, and hopefully get suggestions on how to fix the issue with ScalaCheck, which currently seems to build against the bootstrap jars and not the newly compiled ones (and therefore fails in tests, because it doesn't see its methods like `==>` or `|:` anymore).

What's the recommended approach here without introducing commits with test-failures in the history?

**Comment by paulp** [comment]
Glad you're pursuing this. Is this javap printing things "helpfully" or have you let the quotation marks become part of the name?

``` diff
        : isub
-       : invokevirtual #                  // Method scala/collection/mutable/BitSet.$plus$eq:(I)Lscala/collection/mutable/BitSet;
+       : invokevirtual #                  // Method scala/collection/mutable/BitSet."+=":(I)Lscala/collection/mutable/BitSet;
```

**Comment by soc** [comment]
Interesting ... I'm pretty sure I didn't introduce the quotation marks. Quotation marks would currently end up as `$quot` anyway.

File: src/library/scala/reflect/NameTransformer.scala [hunk 1]
```diff
@@ -71,13 +64,13 @@ object NameTransformer {
...
-        }
-        buf.append("$u%04X".format(c.toInt))
-      }
+//      else if (!Character.isJavaIdentifierPart(c)) {
+//        if (buf eq null) {
+//          buf = new StringBuilder()
+//          buf.append(name.substring(0, i))
+//        }
+//        buf.append("$u%04X".format(c.toInt))
+//      }
```
Comment (soc) [inline src/library/scala/reflect/NameTransformer.scala:73]: Ooops. I wanted to remove this, leaving it as a comment was not my intention.

Comment (JamesIry) [inline src/library/scala/reflect/NameTransformer.scala:73]: Ping. Just reminding that you wanted to whack this comment.

**Comment by JamesIry** [comment]
This is really cool. But it does make these identifiers unreachable from Java. I'm okay with that since "$plus$eq" isn't exactly pleasant to write so libraries targeting Java users already tend to have two APIs. None-the-less this probably edges just over the SIP-worthy bar for that reason alone.

**Comment by paulp** [comment]
Can one of the admins verify this patch?

**Comment by soc** [comment]
There is a discussion on the mailing list: https://groups.google.com/d/topic/scala-internals/7gpyWMg9Bzg/discussion

I think consensus is that more discussion and probably a SIP is needed. I hope people start commenting here on the details of it, but it shouldn't be merged as0is at the moment.

**Comment by JamesIry** [comment]
@soc, would it make sense to close this pr and re-submit if/when any kind of consensus arises around implementing something like this?

**Comment by adriaanm** [comment]
I think that's reasonable.

**Pull request closed without merging**